### PR TITLE
Add transforms for BYOL

### LINKED ISSUE #1324: Clarify SimCLR vs BYOL augmentations
The BYOL paper is a bit misleading as it states that it uses SimCLR augmentations:
image

However, BYOL additionally uses Solarization:
image

At the same time they also made an ablation with using symmetric augmentations or for not using e.g. solarization:
image

The difference without solarize seems to be only `0.2`. With fewer epochs as we use it might be less. We should probably introduce a BYOLTransform that uses the correct augmentations.

### PR DESCRIPTION
**Pull request by IgorSusmelj**
## Closes #1324 

- BYOL augmentations are now separate
- Weaker color jitter as described in the paper
- Asymmetric transformations between the views
- Additional solarization

I'm not very excited about the design as we use just dicts. We don't get any typing. It's the same setup as the original BYOL code uses. But having to make another class just for the parameters sounds also not ideal.

### CODE REVIEW

**Review by guarin — COMMENTED** [review]
I agree that dicts as arguments are not ideal. Did you consider the following setup:

```
class BYOLTransform(MultiViewTransform):
    def __init__(self, view_1: BYOLView1Transform = BYOLView1Transform(), view_2: BYOLView2Transform = BYOLView2Transform()):
        ...

class BYOLView1Transform:
    def __init__(self, ...) # defaults for view 1

class BYOLView2Transform:
    def __init__(self, ...) # defaults for view 2
```

In other parts of the code we have sometimes prefixed arguments like this:
```
global_input_size: int = 224
local_input_size: int = 96
...
```
For example in DINOTransform. But this mostly works if not all arguments can be different. For SMOG we have lists for every argument:
```
input_sizes: [224, 224],
cj_probs: [0.8, 0.8],
...
```
Might also be an option.

**Comment by IgorSusmelj** [comment]
> I agree that dicts as arguments are not ideal. Did you consider the following setup:
> 
> ```
> class BYOLTransform(MultiViewTransform):
>     def __init__(self, view_1: BYOLView1Transform = BYOLView1Transform(), view_2: BYOLView2Transform = BYOLView2Transform()):
>         ...
> 
> class BYOLView1Transform:
>     def __init__(self, ...) # defaults for view 1
> 
> class BYOLView2Transform:
>     def __init__(self, ...) # defaults for view 2
> ```
> 
> In other parts of the code we have sometimes prefixed arguments like this:
> 
> ```
> global_input_size: int = 224
> local_input_size: int = 96
> ...
> ```
> 
> For example in DINOTransform. But this mostly works if not all arguments can be different. For SMOG we have lists for every argument:
> 
> ```
> input_sizes: [224, 224],
> cj_probs: [0.8, 0.8],
> ...
> ```
> 
> Might also be an option.

I did consider it but was not sure as we would have to have two defaults as the Augmentations are asymmetric. But I'll probably go with this then. The other issue I discovered is that with the current dict setup the type hints mess up some of the floats as being integers (e.g. `1.0` turns into `1` and `0.0` turns into `0`)

File: lightly/transforms/byol_transform.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                The input image to apply the transforms to.
+
+        Returns:
+            The transformed image.
+
+        """
+        return self.transform(image)
+
+
+class BYOLTransform(MultiViewTransform):
```
Comment (guarin) [inline lightly/transforms/byol_transform.py:132]: We put the main transform at the top of the file for the other transforms.

File: lightly/transforms/byol_transform.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        - Random gray scale
+        - Gaussian blur
+        - Solarization
+        - ImageNet normalization
+
+    Note that SimCLR v1 and v2 use similar augmentations. In detail, BYOL has
+    asymmetric gaussian blur and solarization. Furthermore, BYOL has weaker
+    color jitter compared to SimCLR.
+
+    - [0]:Bootstrap Your Own Latent, 2020, https://arxiv.org/pdf/2006.07733.pdf
```
Comment (guarin) [inline lightly/transforms/byol_transform.py:154]:
```suggestion
    - [0]: Bootstrap Your Own Latent, 2020, https://arxiv.org/pdf/2006.07733.pdf
```

**Review by guarin — APPROVED** [review]
LGTM!

Left some minor comments but feel free to resolve and merge.

File: lightly/transforms/byol_transform.py (hunk 1 above)
Comment (IgorSusmelj) [reply lightly/transforms/byol_transform.py:132]: I'm not sure whether I can switch the order as we use the `BYOLView1Transform` in the type hint and it has to be initialized before using it.

**Comment by IgorSusmelj** [comment]
Looks like the corrected augmentations could indeed have a slightly impact on the final model accuracy
Red is the model with SImCLR augmentation. Orange is the one with the correct BYOL augmentations.
image

**Review by guarin — APPROVED** [review]
LGTM!

**Pull request merged**